File: src/main/java/pl/training/workshop/m8/s04_shadowlimits/step3/BookingPlan.java
package pl.training.workshop.m8.s04_shadowlimits.step3;

import java.util.List;

import pl.training.workshop.shared.Money;

/**
 * Krok 3: wynik czystego obliczenia - odpowiedź dla klienta i lista efektów DO wykonania.
 * Efekty to dane: można je porównać, zalogować albo wykonać, ale plan sam niczego nie robi.
 */
public record BookingPlan(String result, List<Effect> effects) {
    public sealed interface Effect {
        String describe();

        void applyTo(Effects port);
    }

    public record Charge(String card, Money amount) implements Effect {
        public String describe() {
            return "CHARGE " + card + ": " + amount;
        }

        public void applyTo(Effects port) {
            port.charge(card, amount);
        }
    }

    public record Save(String row) implements Effect {
        public String describe() {
            return "SAVE " + row;
        }

        public void applyTo(Effects port) {
            port.save(row);
        }
    }

    public record SendMail(String to, String text) implements Effect {
        public String describe() {
            return "MAIL " + to + ": " + text;
        }

        public void applyTo(Effects port) {
            port.sendMail(to, text);
        }
    }
}

File: csharp/src/Training.Workshop/M8/S04ShadowLimits/Step3/BookingPlan.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M8.S04ShadowLimits.Step3;

/// <summary>
/// Krok 3: wynik czystego obliczenia - odpowiedź dla klienta i lista efektów DO wykonania.
/// Efekty to dane: można je porównać, zalogować albo wykonać, ale plan sam niczego nie robi.
/// </summary>
public sealed record BookingPlan(string Result, IReadOnlyList<BookingPlan.IEffect> Effects)
{
    public interface IEffect
    {
        string Describe();

        void ApplyTo(IEffects port);
    }

    public sealed record Charge(string Card, Money Amount) : IEffect
    {
        public string Describe()
        {
            return "CHARGE " + Card + ": " + Amount;
        }

        public void ApplyTo(IEffects port)
        {
            port.Charge(Card, Amount);
        }
    }

    public sealed record Save(string Row) : IEffect
    {
        public string Describe()
        {
            return "SAVE " + Row;
        }

        public void ApplyTo(IEffects port)
        {
            port.Save(Row);
        }
    }

    public sealed record SendMail(string To, string Text) : IEffect
    {
        public string Describe()
        {
            return "MAIL " + To + ": " + Text;
        }

        public void ApplyTo(IEffects port)
        {
            port.SendMail(To, Text);
        }
    }
}
